core/src/test/scala/unit/kafka/server/ControllerMutationQuotaTest.scala

```diff
@@ -0,0 +1,361 @@
+  val PermissiveCreateTopicsRequestVersion = 5.toShort
+
+  val StrictDeleteTopicsRequestVersion = ApiKeys.DELETE_TOPICS.latestVersion
+  val PermissiveDeleteTopicsRequestVersion = 4.toShort
+
+  val StrictCreatePartitionsRequestVersion = ApiKeys.CREATE_PARTITIONS.latestVersion
+  val PermissiveCreatePartitionsRequestVersion = 2.toShort
+
+  val TopicsWithOnePartition = Seq("topic-1" ->  1, "topic-2" ->  1)
+  val TopicsWith30Partitions = Seq("topic-1" -> 30, "topic-2" -> 30)
+  val TopicsWith31Partitions = Seq("topic-1" -> 31, "topic-2" -> 31)
+
+  val ControllerMutationRate = 2.0
+}
+
+class ControllerMutationQuotaTest extends BaseRequestTest {
+  import ControllerMutationQuotaTest._
+
+  override def brokerCount: Int = 1
+
+  override def brokerPropertyOverrides(properties: Properties): Unit = {
+    properties.put(KafkaConfig.ControlledShutdownEnableProp, "false")
+    properties.put(KafkaConfig.OffsetsTopicReplicationFactorProp, "1")
+    properties.put(KafkaConfig.OffsetsTopicPartitionsProp, "1")
+    properties.put(KafkaConfig.PrincipalBuilderClassProp,
+      classOf[ControllerMutationQuotaTest.TestPrincipalBuilder].getName)
+    // We use the default number of samples and window size.
+    properties.put(KafkaConfig.NumControllerQuotaSamplesProp, "11")
+    properties.put(KafkaConfig.ControllerQuotaWindowSizeSecondsProp, "1")
+  }
+
+  @Before
+  override def setUp(): Unit = {
+    super.setUp()
+
+    // Define a quota for ThrottledPrincipal
+    defineUserQuota(ThrottledPrincipal.getName, Some(ControllerMutationRate))
+    waitUserQuota(ThrottledPrincipal.getName, ControllerMutationRate)
+  }
+
+  @Test
+  def testSetUnsetQuota(): Unit = {
+    val principal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "User")
+    // Default Value
+    waitUserQuota(principal.getName, Long.MaxValue)
+    // Define a new quota
+    defineUserQuota(principal.getName, Some(ControllerMutationRate))
+    // Check it
+    waitUserQuota(principal.getName, ControllerMutationRate)
+    // Remove it
+    defineUserQuota(principal.getName, None)
+    // Back to the default
+    waitUserQuota(principal.getName, Long.MaxValue)
+  }
+
+  @Test
+  def testStrictCreateTopicsRequest(): Unit = {
+    asPrincipal(ThrottledPrincipal) {
+      // Create two topics worth of 30 partitions each. As we use a strict quota, we
+      // expect the first topic to be created and the second to be rejected.
+      // Theoretically, the throttle time should be bellow or equal to:
+      // ((30 / 10) - 2) / 2 * 10 = 5s
+      val (throttleTimeMs1, errors1) = createTopics(TopicsWith30Partitions, StrictCreateTopicsRequestVersion)
+      assertTrue((5000 - throttleTimeMs1) < 1000)
+      assertEquals(Seq(Errors.NONE, Errors.THROTTLING_QUOTA_EXCEEDED), errors1)
+
+      // The implementation of the Rate has NOT been changed yet so we have to wait past
+      // the window in order to get the avg rate bellow the quota.
+      Thread.sleep(11000) // Thread.sleep(throttleTimeMs1)
```

Comment:
Can we test using mock time or a different approach that doesn't need these long sleeps?

Reply:
I do agree that this is pretty annoying. I am a bit reluctant to using mock time here as these are real integration tests so I would like to test the real conditions. When we will have the improvement for the quota, the sleep time will be <5s and we have 3 of them in the whole test suite. IMO, 15s is not that bad. We do similar in the other quota integration tests for example.

I may be able to reduce it a bit by tweaking the quota used. Let me try.

Reply:
In these type of tests, it is better to use `waitUntil` some condition rather than sleep for a long time. The sleep seems very long because there are multiple tests like this. If it is possible to wait for some condition that typically finishes sooner but may some times take longer, that would be better.

Reply:
That makes sense. I have changed them to using `waitUntilTrue`.